timemachine/fe/rbfe.py

```diff
@@ -122,205 +132,146 @@ def setup_initial_states(st, host_config, temperature, lambda_schedule, seed, mi
+        ligand_conf = st.combine_confs(conf_a, conf_b, lamb)
 
-        assert lamb >= 0.0 and lamb <= 1.0
+        # use a different seed to initialize every window,
+        # but in a way that should be symmetric for
+        # A -> B vs. B -> A edge definitions
+        init_seed = int(seed + hash(ligand_conf.tobytes())) % 10000
 
-        if lamb < 0.5:
-            ligand_conf = st.combine_confs_lhs(mol_a_conf, mol_b_conf)
-        else:
-            ligand_conf = st.combine_confs_rhs(mol_a_conf, mol_b_conf)
-
-        run_seed = seed + lamb_idx
-
-        if host is not None:
-            # run in an environment
-            host_system, host_masses, host_conf = host
-            system = st.combine_with_host(host_system, lamb=lamb)
-            combined_masses = np.concatenate([host_masses, st.combine_masses()])
-            potentials = system.get_U_fns()
-            hmr_masses = model_utils.apply_hmr(combined_masses, system.bond.get_idxs())
-            group_idxs = get_group_indices(get_bond_list(system.bond))
-            baro = MonteCarloBarostat(len(hmr_masses), 1.0, temperature, group_idxs, 15, run_seed + 1)
-            x0 = np.concatenate([host_conf, ligand_conf])
-            box0 = host_config.box
+        if host is None:
+            x0, box0, hmr_masses, potentials, baro = setup_in_vacuum(st, ligand_conf, lamb)
         else:
-            # run a vacuum simulation
-            system = st.setup_intermediate_state(lamb)
-            combined_masses = np.array(st.combine_masses())
-            potentials = system.get_U_fns()
-            hmr_masses = model_utils.apply_hmr(combined_masses, system.bond.get_idxs())
-            baro = None
-            x0 = ligand_conf
-            box0 = np.eye(3, dtype=np.float64) * 10  # make a large 10x10x10nm box
-
-        # minimize water box around the ligand by 4D-decoupling
-
-        # hmr masses should be identical throughout the lambda schedule
-        # bond idxs should be the same at the two end-states, note that a possible corner
-        # case with bond breaking may seem to be problematic:
-
-        # 0 1 2    0 1 2
-        # C-O-C -> C.H-C
-
-        # but this isn't an issue, since hydrogens will only ever be terminal atoms
-        # and core hydrogens that are mapped to heavy atoms will take the mass of the
-        # heavy atom (thereby not triggering the mass repartitioning to begin with).
+            x0, box0, hmr_masses, potentials, baro = setup_in_env(
+                st, host, host_config, ligand_conf, lamb, temperature, init_seed
+            )
 
-        # but its reasonable to be skeptical, so we also assert consistency through the lambda
-        # schedule as an extra sanity check.
+        # provide a different run_seed for every lambda window,
+        # but in a way that should be symmetric for
+        # A -> B vs. B -> A edge definitions
+        run_seed = int(seed + hash(bytes().join([np.array(p.params).tobytes() for p in potentials]))) % 10000
```

Comment:
Have you checked if symmetry is preserved for the x0 when flipping the direction?

Reply:
Yes, confirmed on a simple case.

Should possibly extract this into a utility function as suggested above https://github.com/proteneer/timemachine/pull/936#discussion_r1092302139 to be able to add a unit test.

Note, the proposed behavior of `init_seed` and `run_seed` is different.

* `init_seed` is a deterministic function of `x0` and `seed`. So currently the same `init_seed` will be generated for all windows where lamb < 0.5, and a second will be generated for all windows where lamb > 0.5. This may change if `x0` is interpolated as a function of lamb.

* `run_seed` is a deterministic function of `potentials`' `params` and `seed`, so a unique `run_seed` will be generated for each window.